# Day 1 : Chronal Calibration

Oh good. The theme this year is temporal anamolies. :)

Here's a nice easy one to get the advent calendar started. I've found the difficulty can vary wildly from day to day so there's that. :)

For this one I need an accumulator, and to parse my input ...

In [17]:
with open('AdventOfCode/input/2018-1.txt') as f:
    frequency_changes = [int(line.strip()) for line in f]
print(frequency_changes[:10])


[-6, -1, -18, -10, 5, 18, 19, 11, -6, 13]


In [21]:
frequency = 0
for c in frequency_changes:
    frequency += c
frequency

538

That's the right answer for part 1.

# Day 1 : Part 2

Now I should keep track of frequencies I've seen and stop when I see a repeated value. (This could request multiple applications of my frequency_changes list.)

In [26]:
seen = {0: True}
frequency = i = cnt = 0
while True:
    frequency += frequency_changes[i]
    i = (i + 1) % len(frequency_changes)
    if frequency in seen: break
    else: seen[frequency] = True
    cnt += 1
print(frequency, cnt)

77271 145416


# Day 2 : Part 1

Today the fabric for Santa's suit has gone missing! We have to look for more in a warehouse with boxes labeled with obscure lettering ... :)

I should keep separate counts of labels with 2 repeated chars, or 3. (Exactly mind you, 1 label might belong in both groups). I'll use these counters next to calculate a checksum ...

In [5]:
def Input(n):
    with open('input/{}.txt'.format(n)) as f:
        return list(line.strip() for line in f)
labels = Input(2)
labels[:10]

['rmyxgdlihczskunpfwbgqoeybv',
 'rmyxgdlksczskunpfwbjqkeatv',
 'rmybgdxibczskunpfwbjqoeatv',
 'rmyxgdlirczskuopfwbjqzeatv',
 'rmyxedlrhczskunpfwbyqoeatv',
 'rmyxfdlicczskunpfwbxqoeatv',
 'rmyxgvlihkzskunpfwbsqoeatv',
 'rmyxgdaihczvkunpfwblqoeatv',
 'nmyxgolihczskunpfwbjqieatv',
 'rhyxgdcihczskunifwbjqoeatv']

In [7]:
import collections

two_of_a_kind = three_of_a_kind = 0
for l in labels:
    c = collections.Counter(l)
    if 2 in c.values(): two_of_a_kind += 1
    if 3 in c.values(): three_of_a_kind += 1
two_of_a_kind * three_of_a_kind

6370

# Day 2 : Part 2

Find 2 labels that differ in only 1 character. Eg 'abcd' and 'adcd' differ in their second character. All others are the same.

In [10]:
import itertools

for s1, s2 in itertools.combinations(labels, 2):
    x = [i for i in range(len(s1)) if s1[i] != s2[i]]
    if len(x) == 1: break
s1, s2, x

('rmyxgdlihczskunpfiyjqcebtv', 'rmyxgdlihczskunpfibjqcebtv', [18])

In [16]:
s1[:x[0]]+s1[x[0]+1:]


'rmyxgdlihczskunpfijqcebtv'

# Day 3 : No Matter How You Slice It

I guess I should keep track of fabric inches and claims

In [45]:
import re
import collections

fabric_inches = collections.defaultdict(list)

f = open('input/3.txt')
for line in f.readlines():
    claim, x, y, w, h = map(int, re.findall('\d+', line))
    for inche in claim_inches(x, y, w, h):
        fabric_inches[inche] += [claim]


In [44]:
def claim_inches(x, y, w, h):
    '''Returns the list of fabric inches covered by a claim'''
    return [(i, j) for i in range(x, x + w)
                   for j in range(y, y + h)]
claim_inches(1,1,4,3)

[(1, 1),
 (1, 2),
 (1, 3),
 (2, 1),
 (2, 2),
 (2, 3),
 (3, 1),
 (3, 2),
 (3, 3),
 (4, 1),
 (4, 2),
 (4, 3)]

That looks ok

In [46]:
a = [1 for i in fabric_inches.values() if len(i) > 1]
len(a)

109143

Guesses
1386 is too low :/
109143 was a better answer lol !

# Part 2

Apparently 1 claim doesn't overlap any other ... all inche claim lists would be length 1
Collect all claim_ids from inches with overlapping ones
This should be all claims but one

In [64]:
from itertools import chain
a = set(range(1, 1283))
overlapping = [l for l in fabric_inches.values() if len(l) > 1]
a-set(chain.from_iterable(overlapping))

{506}